File: 1930.unique-length-3-palindromic-subsequences.py
#
# @lc app=leetcode id=1930 lang=python3
#
# [1930] Unique Length-3 Palindromic Subsequences
#

# @lc code=start
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        # Find number of subsequences in s
        # Subsequence must be length 3 and must be unique palindromes
        ans = set()
        after = [0] * 26
        pre = [0] * 26

        n = len(s)
        for i in range(1, n):
            after[ord(s[i]) - ord('a')] += 1
        
        for i in range(1, n-1):
            pre[ord(s[i-1])- ord('a')] += 1
            after[ord(s[i]) - ord('a')] -= 1

            for j in range(26 ):
                if pre[j] > 0 and after[j] > 0:
                    cur = chr(j + ord("a")) + s[i] + chr(j + ord("a"))
                    ans.add(cur )
        return len(ans)
# @lc code=end

File: 1930.unique-length-3-palindromic-subsequences.java
/*
 * @lc app=leetcode id=1930 lang=java
 *
 * [1930] Unique Length-3 Palindromic Subsequences
 */

// @lc code=start
import java.util.HashSet;
import java.util.Set;
class Solution {
    public int countPalindromicSubsequence(String s) {
        int ans = 0;
        Set<Character> Uniq = new HashSet<>();
        for (char c : s.toCharArray()){
            Uniq.add(c);
        }
        for(char c : Uniq){
            int first = s.indexOf(c);
            int last = s.lastIndexOf(c);
            if (first < last) {
                Set<Character> middleChars = new HashSet<>();
                for (int i = first + 1; i < last; i++) {
                    middleChars.add(s.charAt(i));
                }
                ans += middleChars.size();
            }
        }
        return ans;

    }
}
// @lc code=end

